In [55]:
import numpy as np
import pandas as pd

from typing import Union

In [56]:
sample = pd.read_csv('train.csv')

In [57]:
sample

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [58]:
sample.shape
#891 запись, 12 переменных

(891, 12)

In [59]:
sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [60]:
sample.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [70]:
sample.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Анализ переменной PassengerId

In [73]:
sample["PassengerId"].describe()

,PassengerId
count,891.000000
mean,446.000000
std,257.353842
min,1.000000
25%,223.500000
50%,446.000000
75%,668.500000
max,891.000000


In [74]:
sample["PassengerId"].nunique()

891

Анализ переменной Survived

In [75]:
sample["Survived"].describe()

,Survived
count,891.000000
mean,0.383838
std,0.486592
min,0.000000
25%,0.000000
50%,0.000000
75%,1.000000
max,1.000000


In [78]:
sample["Survived"].nunique()

2

Анализ переменной Pclass

In [80]:
sample["Pclass"].describe()

,Pclass
count,891.000000
mean,2.308642
std,0.836071
min,1.000000
25%,2.000000
50%,3.000000
75%,3.000000
max,3.000000


In [81]:
sample["Pclass"].nunique()

3

Анализ переменной Name

In [85]:
sample["Name"].describe()

,Name
count,891
unique,891
top,"Braund, Mr. Owen Harris"
freq,1


Анализ переменной Sex

In [86]:
sample["Sex"].describe()

,Sex
count,891
unique,2
top,male
freq,577


Анализ переменной Age

In [89]:
sample["Age"].describe()

,Age
count,714.000000
mean,29.699118
std,14.526497
min,0.420000
25%,20.125000
50%,28.000000
75%,38.000000
max,80.000000


In [90]:
sample["Age"].nunique()

88

Анализ переменной SibSp

In [91]:
sample["SibSp"].describe()

,SibSp
count,891.000000
mean,0.523008
std,1.102743
min,0.000000
25%,0.000000
50%,0.000000
75%,1.000000
max,8.000000


In [92]:
sample["SibSp"].nunique()

7

Анализ переменной Parch

In [98]:
sample["Parch"].describe()

,Parch
count,891.000000
mean,0.381594
std,0.806057
min,0.000000
25%,0.000000
50%,0.000000
75%,0.000000
max,6.000000


In [99]:
sample["Parch"].nunique()

7

Анализ переменной Ticket

In [103]:
sample["Ticket"].describe()

,Ticket
count,891
unique,681
top,347082
freq,7


Анализ переменной Fare

In [104]:
sample["Fare"].describe()

,Fare
count,891.000000
mean,32.204208
std,49.693429
min,0.000000
25%,7.910400
50%,14.454200
75%,31.000000
max,512.329200


In [105]:
sample["Fare"].nunique()

248

Анализ переменной Cabin

In [109]:
sample["Cabin"].describe()

,Cabin
count,204
unique,147
top,B96 B98
freq,4


Анализ переменной Embarked

In [110]:
sample["Embarked"].describe()

,Embarked
count,889
unique,3
top,S
freq,644


In [111]:
sample["Embarked"].nunique()

3

In [63]:
#класс для анализа датасета

class analysis_dataset:
  def __init__(self,sample):
    self.sample = sample

  #вывод n первых (order = head) и последних (order = tail) записей таблицы
  #аналоги head и tail из pandas
  def show(self, order: str, num: int) -> list:
    if num <= len(self.sample):

      if order == "head":
        return self.sample[0:num]
      elif order == "tail":
        return self.sample[len(self.sample)-num-1:len(self.sample)-1]
      else:
        print("Выбран неверный порядок! Укажите head или tail!")

    else:
      print("Количество выводимых символов превышает величину таблицы!")

  #вывод основных статистик descibe() для переменной
  def describe_of_variable(self, var1: str) -> list | None:
    if var1 not in self.sample.columns:
      print(f"Указанных переменных нет в анализируемом сэмпле!")
    else:
      return self.sample[var1].describe()

  #вывод статистик по двум переменным
  def table_with_stats(self,var1: str,var2: str) -> list | None:
    if var1 not in self.sample.columns or var2 not in self.sample.columns:
      print(f"Указанных переменных нет в анализируемом сэмпле!")
    else:
      return self.sample[[var1,var2]].groupby(var2).agg(["count","mean"])

  #вывод коэффициента корреляции между двумя переменными
  def correlation(self,var1: str,var2: str) -> list | None:
    if var1 not in self.sample.columns or var2 not in self.sample.columns:
      print(f"Указанных переменных нет в анализируемом сэмпле!")
    else:
      return self.sample[var1].corr(self.sample[var2])

In [64]:
#создание объекта класса
sample1 = analysis_dataset(pd.read_csv('train.csv'))
#первые 10 значений
sample1.show("head",10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [40]:
#последние 5 значений
sample1.show("tail",5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.125,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.450,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.000,C148,C


In [41]:
sample1.show("tall",5)

Выбран неверный порядок! Укажите head или tail!


In [42]:
sample1.show("tail",1000)

Количество выводимых символов превышает величину таблицы!


Анализ переменной Pclass

In [82]:
sample1.describe_of_variable("Pclass")

,Pclass
count,891.000000
mean,2.308642
std,0.836071
min,1.000000
25%,2.000000
50%,3.000000
75%,3.000000
max,3.000000


In [83]:
sample1.table_with_stats("Survived", "Pclass")

Survived          
          count      mean
Pclass                   
1           216  0.629630
2           184  0.472826
3           491  0.242363

In [84]:
sample1.correlation("Survived", "Pclass")

-0.3384810359610148

Из анализа выше видно, что существует достаточно сильная зависимость между классами переменной Pclass и вероятностью выживания: 63% для Pclass = 1, 47% для Pclass = 2, 24% для Pclass = 3

Анализ переменной Sex

In [87]:
sample1.describe_of_variable("Sex")

,Sex
count,891
unique,2
top,male
freq,577


In [88]:
sample1.table_with_stats("Survived","Sex")

Survived          
          count      mean
Sex                      
female      314  0.742038
male        577  0.188908

Вероятность выживания среди женщин (74%) существенно выше, чем среди мужчин (19%)

Анализ переменной Age

In [52]:
sample1.describe_of_variable("Age")

,Age
count,714.000000
mean,29.699118
std,14.526497
min,0.420000
25%,20.125000
50%,28.000000
75%,38.000000
max,80.000000


In [53]:
sample1.table_with_stats("Survived","Age")

Survived     
         count mean
Age                
0.42         1  1.0
0.67         1  1.0
0.75         2  1.0
0.83         2  1.0
0.92         1  1.0
...        ...  ...
70.00        2  0.0
70.50        1  0.0
71.00        2  0.0
74.00        1  0.0
80.00        1  1.0

[88 rows x 2 columns]

In [54]:
sample1.correlation("Survived","Age")

-0.07722109457217768

Анализ переменной SibSp

In [95]:
sample1.describe_of_variable("SibSp")

,SibSp
count,891.000000
mean,0.523008
std,1.102743
min,0.000000
25%,0.000000
50%,0.000000
75%,1.000000
max,8.000000


In [96]:
sample1.table_with_stats("Survived","SibSp")

Survived          
         count      mean
SibSp                   
0          608  0.345395
1          209  0.535885
2           28  0.464286
3           16  0.250000
4           18  0.166667
5            5  0.000000
8            7  0.000000

In [97]:
sample1.correlation("Survived","SibSp")

-0.03532249888573564

Вероятность выживания для SibSP = 0 (35%) существенно ниже, чем SibSP = 1 (54%) или 2 (46%)

Анализ переменной Parch

In [100]:
sample1.describe_of_variable("Parch")

,Parch
count,891.000000
mean,0.381594
std,0.806057
min,0.000000
25%,0.000000
50%,0.000000
75%,0.000000
max,6.000000


In [101]:
sample1.table_with_stats("Survived","Parch")

Survived          
         count      mean
Parch                   
0          678  0.343658
1          118  0.550847
2           80  0.500000
3            5  0.600000
4            4  0.000000
5            5  0.200000
6            1  0.000000

In [102]:
sample1.correlation("Survived","Parch")

0.08162940708348371

Вероятность выживания для Parch = 0 (35%) существенно ниже, чем Parch > 0 (50%)

Анализ переменной Fare

In [106]:
sample1.describe_of_variable("Fare")

,Fare
count,891.000000
mean,32.204208
std,49.693429
min,0.000000
25%,7.910400
50%,14.454200
75%,31.000000
max,512.329200


In [107]:
sample1.table_with_stats("Survived","Fare")

Survived          
            count      mean
Fare                       
0.0000         15  0.066667
4.0125          1  0.000000
5.0000          1  0.000000
6.2375          1  0.000000
6.4375          1  0.000000
...           ...       ...
227.5250        4  0.750000
247.5208        2  0.500000
262.3750        2  1.000000
263.0000        4  0.500000
512.3292        3  1.000000

[248 rows x 2 columns]

In [108]:
sample1.correlation("Survived","Fare")

0.2573065223849625

Есть корелляция между Fare и показателем выживания (>25%)

Анализ переменной Embarked

In [112]:
sample1.describe_of_variable("Embarked")

,Embarked
count,889
unique,3
top,S
freq,644


In [113]:
sample1.table_with_stats("Survived","Embarked")

Survived          
            count      mean
Embarked                   
C             168  0.553571
Q              77  0.389610
S             644  0.336957

Вероятность выживания для Embarked = C (55%) существенно выше, чем для Embarked = Q (39%) и S (34%)

В процессе анализа была обнаружено влияние переменных  Pclass, Sex, SibSp, Parch, Embarked (различия между классами переменных в вероятности выживания) и Fare (высокая корреляция) на целевую переменную Survived.